### ISSUE TITLE
Support for 4k (Photon Mono X)

### ISSUE THREAD

**Issue by mperino** [fookatchu/SL1toPhoton#12]
Hello, It looks like photonfile.py is hard coded to 1440x2560.   I see some good checking logic, and most of your code seems hard coded for 1440x2560. 

When testing files in anticipation of using the Photon Mono X which is 3840x2400.  I dont have one yet, but looking at the slicer options, I would greatly prefer to use sl1tophoton as my preferred method :) 

Is support for 4K something that you are already working on?

Nice easy to read code by the way. If there is a branch that you are working on for 4K please let me know. 

Would you consider adding a printerconf file, perhaps using configparser:
[DEFAULT]
xpixel= 1440
ypixel = 2560
zheight = 155
antialias = NO

[Photon Mono X]
xpixel= 3840
ypixel = 2400
zheight = 240
antialias = NO

Then adding a selector to the GUI, and --option to select the printer?

**Comment by cab404** [comment]
Hi! I am currently working on a fork of SL1ToPhoton. Can you please attach an example .photon file?

**Comment by mperino** [comment]
Hi Cab404, 

I have an issue open with AnyCubic:
https://github.com/ANYCUBIC-3D/PhotonWorkshop/issues/9

for exactly that. It's clear from some of the Photon Mono X videos out there, that 4k is supported but the selector doesnt allow it by default.  So we need AnyCubic or someone to tell us how they got it enabled in Photon WorkShop.

I've seen both your and X3msnake's posts here, and hoping that between all of us and fookatchu we make this work :)

**Comment by cab404** [comment]
> I've seen both your and X3msnake's posts here, and hoping that between all of us and fookatchu we make this work :)

It's not hard to export higher resolution images — pyphotonfile is really flexible.

Also, you don't really need additional configs for that — printer information, including resolution and size is included into sl1 files.

**Comment by cab404** [comment]
> Hi Cab404,
> 
> I have an issue open with AnyCubic:
> [ANYCUBIC-3D/PhotonWorkshop#9](https://github.com/ANYCUBIC-3D/PhotonWorkshop/issues/9)
> 
> for exactly that. It's clear from some of the Photon Mono X videos out there, that 4k is supported but the selector doesnt allow it by default. So we need AnyCubic or someone to tell us how they got it enabled in Photon WorkShop.
> 
> I've seen both your and X3msnake's posts here, and hoping that between all of us and fookatchu we make this work :)

BTW, can you please send me a sliced `.photon` file from the latest PhotonWorkshop? It's quite a hassle to get it running on my platform)

**Comment by mperino** [comment]
I've got it running nicely under Wine 5.0-3.  2 versions currently installed.  V2.1.21RC6 (was latest git last week), and 2.1.20RC1 (version currently being served by their website).

I have a test .stl that is a 1x1x1 Cylinder.
I sliced that with both versions and outputted it for the default generic Photon with no changes to any settings.
Files are in my fork under the examples folder:
https://github.com/mperino/pyphotonfile/tree/master/examples

Once I get photon workshop working at the higher resolutions, I'll post an example at 4k of the same .stl (1x1x1), and scale the same file up in the X&Y to the largest that fits the bed.

**Comment by cab404** [comment]
Wine is fussy on unstable NixOS right now :\

Thanks, you really helped getting defaults for v2 right)

You are sure these are default settings, right? (bottom lifting speed seems dangerously high)

**Comment by mperino** [comment]
![image](https://user-images.githubusercontent.com/3091956/93135516-4e249d00-f68f-11ea-8ea2-a05d99ed6100.png)

That's what both show in the GUI.

**Comment by cab404** [comment]
> ![image](https://user-images.githubusercontent.com/3091956/93135516-4e249d00-f68f-11ea-8ea2-a05d99ed6100.png)
> 
> That's what both show in the GUI.

Can you also slice a thing that requires supports? I want to see, if that would change lifting times~

**Comment by mperino** [comment]
I added one to my fork:
https://github.com/mperino/pyphotonfile/tree/master/examples

sm Cross Piece x2.photon  is sliced with native supports (not exported from prusaslicer).
I also Included a screenshot of the slice settings and model (still slicing using 1440x2560).
![image](https://user-images.githubusercontent.com/3091956/93691835-e95aaf80-fa9f-11ea-8f35-20f23b660293.png)

**Comment by mperino** [comment]
Ok, after digging a lot deeper and starting to understand how SL1toPhoton and pyphotonfile mesh together (mostly what does what).

It looks to me like the sanest thing could be to have a method for SL1toPhoton to use:
display_height = 2400
display_orientation = portrait
display_pixels_x = 508
display_pixels_y = 508
display_width = 3840

Seems like it wouldnt be a bad thing to trust the .sl1 file for a bunch of settings that currently are discarded?

photon.bed_x(y,z) already exists in the write out for pyphotonfile, so if we trust the values from the .sl1 file, it'd be a cleanish way to go (admittedly havent looked at what this breaks)..

the checking and reshape would need to change from hard coded to the config variable:  rgb2d = x.reshape((2560,1440)) should be: rgb2d = x.reshape((bed_x,bed_y))

Am I way off, before I spend more time in that direction?

**Comment by cab404** [comment]
Nope, you are quite right. If we are to generalize SL1ToPhoton to more printers, we would need to read prusaslicer.ini from sl1 file, with all the printer data (copypaste [this](https://github.com/cab404/SL1toPhoton/blob/master/SL1_to_Photon.py#L24) method, change to prusaslicer.ini).

Although I didn't try exporting different sizes and resolutions, mainly because I don't have means to test it :(

**Comment by mperino** [comment]
I don't have any method to test em yet either.. Hoping to get my Mono X between Oct 24 and Nov 7th.

My fork:

https://github.com/mperino/SL1toPhoton/blob/master/SL1_to_Photon.py  

Has logic to read the bed size from the prusaslicer.ini (took me forever to realize that it was a seperate file (DOH!).  Currently the logic is to die if it cant read the bed size, but easily changeable to default to 1440
the try clause bellow is for importing settings from the prusaslicer.ini in the .sl1 file.
Current logic is to bomb out if we cant find the bed size, but we could default back to 1440 by uncommenting and commenting out the exit

try:
        photon.bed_x = int(sl1.prusaslicer['display_width'])
        photon.bed_y = int(sl1.prusaslicer['display_height'])
except:
       print('ERROR: Unable to set display width or height from .sl1 file')
       # print ("setting a default size of 1440x2560")
       # photon.bed_x = 1440
       # photon.bed_y = 2560
       sys.exit(-1)


Mine is based off fookachu's main, but Im going to see how many more of your hard coded settings I can pull from the .sl1 file before diving into pyphotonfile to make it aware of variable bed size. A few other changes as well for things that were griping about older syntax.  I'd like to get as much of it based off settings in prusaslicer, and as little as possible hard coded.  

Still no response from ANYCubic on how to enable 4K in their slicer...

**Comment by cab404** [comment]
I added template importing from SL1 and multithreaded slicing, and fixed some mistakes in pyphotonfile

**Comment by mperino** [comment]
@cab404 I finally got a copy of photon workshop that slices Photon Mono X.  You'll want version 2.1.23 for support of Mono X (and the other new ones). 

Even faster default speeds and shorter exposure times:
![image](https://user-images.githubusercontent.com/3091956/95391803-8a24d980-08ac-11eb-941f-0805ea46328e.png)

Link to 2.1.23:
https://drive.google.com/file/d/12YNF_7KfwSULl9JAFw_CF9DU0eGeciA0/view?usp=sharing
